src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,380 @@
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Runtime.ExceptionServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Http
+{
+    /// <summary>
+    /// Implements PipeReader using an underlying stream.
+    /// </summary>
+    public class StreamPipeReader : PipeReader
+    {
+        private readonly int _minimumSegmentSize;
+        private readonly Stream _readingStream;
+        private readonly MemoryPool<byte> _pool;
+
+        private CancellationTokenSource _internalTokenSource;
+        private bool _isCompleted;
+        private ExceptionDispatchInfo _exceptionInfo;
+
+        private BufferSegment _readHead;
+        private int _readIndex;
+
+        private BufferSegment _readTail;
+        private long _consumedLength;
+        private bool _examinedEverything;
+
+        private CancellationTokenSource InternalTokenSource
+        {
+            get => _internalTokenSource
+                            ?? Interlocked.CompareExchange(ref _internalTokenSource, new CancellationTokenSource(), null)
+                            ?? _internalTokenSource;
+        }
+
+        /// <summary>
+        /// Creates a new StreamPipeReader.
+        /// </summary>
+        /// <param name="readingStream">The stream to read from.</param>
+        public StreamPipeReader(Stream readingStream) : this(readingStream, minimumSegmentSize: 4096)
+        {
+        }
+
+        /// <summary>
+        /// Creates a new StreamPipeReader.
+        /// </summary>
+        /// <param name="readingStream">The stream to read from.</param>
+        /// <param name="minimumSegmentSize">The minimum segment size to return from ReadAsync.</param>
+        /// <param name="pool"></param>
+        public StreamPipeReader(Stream readingStream, int minimumSegmentSize, MemoryPool<byte> pool = null)
+        {
+            _minimumSegmentSize = minimumSegmentSize;
+            _readingStream = readingStream;
+            _pool = pool ?? MemoryPool<byte>.Shared;
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            ThrowIfCompleted();
+
+            if (_readHead == null || _readTail == null)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoDataRead();
+            }
+            // Creating consumedSequence will throw an ArgumentOutOfRangeException if consumed/examined are invalid.
+            // We want the same check for netstandard too.
+            // TODO this is significantly slower than just calling AdvanceTo().
+            var consumedSequence = GetCurrentReadOnlySequence().Slice(consumed, examined);
+
+#if NETCOREAPP2_2
```

Comment:
Why do we have these checks? Why bother using TryGetReadOnlySequenceSegment? cc @pakrym ?

Reply:
It causes a pretty bit perf hit on AdvanceTo.

Reply:
Yes I think we should remove this and directly use the cursor (I'll fight with @pakrym later 😉 )